mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -373,11 +385,57 @@ static bool isUnfusedKernelStore(ThreadwiseWriteAllOp store) {
   return ret;
 }
 
+static SmallVector<Attribute> extractVector(ArrayAttr arrayAttr) {
+  return llvm::to_vector<4>(arrayAttr.getAsRange<Attribute>());
+}
+
+LogicalResult
+ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+                                      PatternRewriter &rewriter) const {
+  Location loc = reduceOp.getLoc();
+  ThreadwiseWriteAllOp threadwiseWriteOp =
+      traceToThreadwiseWrite(reduceOp.getIn());
+  if (reduceOp.getReduceMethod() != ReduceMethod::Sum) {
+    return reduceOp.emitError("We only support sum reductions.!");
+  }
+  if (!threadwiseWriteOp) {
+    LLVM_DEBUG(llvm::dbgs() << "rock.reduce input is not leading to "
+                               "ThreadwiseWriteAllOp of the previous op.");
+    return failure();
+  }
+  int64_t reductionAxis = reduceOp.getAxisAttr().getInt();
+  TypedValue<ShapedType> redIn = reduceOp.getIn();
+  ArrayRef<int64_t> reduceInShape = redIn.getType().getShape();
+  TopDownTMBuilder dropReductionDim(rewriter, reduceInShape, loc);
+  for (uint32_t i = 0; i < reduceInShape.size(); ++i) {
+    if (i == reductionAxis) {
+      dropReductionDim.constDim("reduction_dim", i, 0, 1);
+    } else {
+      dropReductionDim.passThrough({i}, {i});
+    }
+  }
+  TransformMapAttr trAttr = dropReductionDim.get();
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(threadwiseWriteOp);
+    redIn = rewriter.create<TransformOp>(loc, redIn, trAttr);
+    threadwiseWriteOp.getDestMutable().assign(reduceOp.getOut());
+    threadwiseWriteOp.setStoreMethodAttr(
+        StoreMethodAttr::get(rewriter.getContext(), StoreMethod::AtomicAdd));
+    auto extraViews = extractVector(threadwiseWriteOp.getExtraViews());
```

Comment:
I'm not sure this is correct. `extraViews` are applied before the transformation chain on the `Value`. You may want to transform `reduceOp.getOut()`

Reply:
Can you elaborate a bit more on the difference?
i.e. taking an extraView on output memref vs transforming output memref and passing that in..

Reply:
Strictly speaking, there's no difference. Threadwise_write_all writes to `outputCoords := viewsOnArgument(extraViews(bid, tid, iter))`

Conceptually, threadwise_write_all always takes an argument that's in GEMM coordinates, and I want to keep it that way.